From 61dec4b9c9b145dd5a7ae4e60607b982b2ac83a5 Mon Sep 17 00:00:00 2001
From: Amit Kumar Mahapatra <amit.kumar-mahapatra@xilinx.com>
Date: Sun, 19 Jan 2020 10:58:28 -0700
Subject: [PATCH 0918/1851] spi: Add multi-die info in spi_device structure

commit adf199e79760dcad21a3470385b512509ce28039 from
https://github.com/Xilinx/linux-xlnx.git

Added multi-die member in spi_device structure.
Scan for multi-die property in the device tree, if
present then set the multi_die member of spi_device
structure variable to true.

If the multi_die member of spi_device structure variable
is set, then a read transaction that spans across multiple
banks is split into one read per bank. This resolves the
die crossover issue.

Signed-off-by: Amit Kumar Mahapatra <amit.kumar-mahapatra@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/spi/spi.c       | 3 +++
 include/linux/spi/spi.h | 2 ++
 2 files changed, 5 insertions(+)

diff --git a/drivers/spi/spi.c b/drivers/spi/spi.c
index fce4a37515da..96b6184300e1 100644
--- a/drivers/spi/spi.c
+++ b/drivers/spi/spi.c
@@ -1773,6 +1773,9 @@ static int of_spi_parse_dt(struct spi_controller *ctlr, struct spi_device *spi,
 	}
 	spi->max_speed_hz = value;
 
+	/* Multi die flash */
+	if (of_property_read_bool(nc, "multi-die"))
+		spi->multi_die = true;
 	return 0;
 }
 
diff --git a/include/linux/spi/spi.h b/include/linux/spi/spi.h
index f87af41c676d..73deefeb33f6 100644
--- a/include/linux/spi/spi.h
+++ b/include/linux/spi/spi.h
@@ -124,6 +124,7 @@ void spi_statistics_add_transfer_stats(struct spi_statistics *stats,
  *	not using a GPIO line)
  * @word_delay_usecs: microsecond delay to be inserted between consecutive
  *	words of a transfer
+ * @multi_die: Flash device with multiple dies.
  *
  * @statistics: statistics for the spi_device
  *
@@ -172,6 +173,7 @@ struct spi_device {
 	int			cs_gpio;	/* LEGACY: chip select gpio */
 	struct gpio_desc	*cs_gpiod;	/* chip select gpio desc */
 	uint8_t			word_delay_usecs; /* inter-word delay */
+	bool			multi_die;	/* flash with multiple dies*/
 
 	/* the statistics */
 	struct spi_statistics	statistics;
-- 
2.31.1

